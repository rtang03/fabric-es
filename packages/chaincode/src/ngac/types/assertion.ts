export interface Assertion {
  sid: string;
  assertion: boolean;
  message?: string;
}
