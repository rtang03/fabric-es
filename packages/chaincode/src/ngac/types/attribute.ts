export interface Attribute {
  type: '1' | 'N';
  key: string;
  value: string | string[];
  alias?: string;
  disabled?: boolean;
  immutable?: boolean;
}
