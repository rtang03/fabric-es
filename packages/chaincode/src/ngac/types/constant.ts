export enum CONTEXT {
  ISSUER_ORGNAME = 'issuer_orgname',
  ISSUER_CN = 'issuer_cn',
  SUBJECT_OUNAME = 'subject_ouname',
  SUBJECT_ORGNAME = 'subject_orgname',
  SUBJECT_CN = 'subject_cn',
  INVOKER_ID = 'invoker_id',
  INVOKER_MSPID = 'invoker_mspid'
}

export enum RESOURCE {
  ENTITYNAME = 'entityName',
  VERSION = 'version',
  CREATOR_MSPID = 'creator_mspid',
  CREATOR_CN = 'creator_cn',
  CREATOR_ID = 'creator_id'
}

export enum NAMESPACE {
  MODEL = 'model',
  ORG = 'org',
  ENTITY = 'entity',
  MSP_ATTRIBUTE = 'mspattr',
  ENTITY_ATTRIBUTE = 'entityattr',
  RESOURCE_ATTRIBUTE = 'resattr',
  POLICY = 'policy'
}
