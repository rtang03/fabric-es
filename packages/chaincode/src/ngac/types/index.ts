export * from './resource';
export * from './action';
export * from './policy';
export * from './attribute';
export * from './constant';
export * from './ngacRepo';
export * from './assertion';
