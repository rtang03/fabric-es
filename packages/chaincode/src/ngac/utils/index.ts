export * from './makeKey';
export * from './createResource';
export * from './createPolicy';
export * from './createId';
export * from './hasList';
export * from './stringEquals';
