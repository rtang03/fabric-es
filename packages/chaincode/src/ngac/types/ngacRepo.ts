import { Attribute } from './attribute';
import { Policy } from './policy';
import { Resource } from './resource';

export interface NgacRepo {
  // add one policy
  addPolicy: (policy: Policy) => Promise<Policy>;
  // add and replace one attribute group to a MSPID
  addMSPAttr: (resource: Resource) => Promise<Attribute[]>;
  // add and replace one attribute group to a reource URI
  addResourceAttr: (resource: Resource) => Promise<Attribute[]>;
  deleteMSPAttrByMSPID: (mspid: string) => Promise<string>;
  // delete all policies by id
  deletePolicyById: (id: string) => Promise<string[]>;
  // delete one policy by id
  deletePolicyByIdSid: (id: string, sid: string) => Promise<string>;
  // delete attribute group by URI
  deleteReourceAttrByURI: (uri: string) => Promise<string>;
  // delete attribute group by MSPID
  getMSPAttrByMSPID: (mspid: string) => Promise<Attribute[]>;
  // get policies by id
  getPolicyById: (id: string) => Promise<Policy[]>;
  // get one policy
  getPolicyByIdSid: (id: string, sid: string) => Promise<Policy>;
  // get array of attribute group by URI path segment
  // this API may be useless; no such use case. This is not exposed as chaincode interface
  getResourceAttrGroupByURI: (uri: string) => Promise<Attribute[][]>;
  // get attribte gropu by URI
  getResourceAttrByURI: (uri: string) => Promise<Attribute[]>;
  // update and insert the attribute group, by appending the existing one
  upsertResourceAttr: (resource: Resource) => Promise<Attribute[]>;
}
