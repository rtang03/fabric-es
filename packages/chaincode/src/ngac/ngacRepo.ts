import { Context } from 'fabric-contract-api';
import { assign, values } from 'lodash';
import stateList from './ledger-api/statelist';
import { Attribute, NAMESPACE as NS, NgacRepo, Policy } from './types';
import { makeKey } from './utils';

export const ngacRepo: (context: Context) => NgacRepo = context => ({
  addMSPAttr: async resource =>
    resource
      ? stateList<Attribute[]>(NS.MSP_ATTRIBUTE, context).addState(
          resource.key,
          resource.mspAttrs
        )
      : null,
  addResourceAttr: async resource =>
    resource
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).addState(
          resource.key,
          resource.resourceAttrs
        )
      : null,
  upsertResourceAttr: async resource => {
    if (!resource) return null;
    const attributes = await stateList<Attribute[]>(
      NS.RESOURCE_ATTRIBUTE,
      context
    ).getState(resource.key);

    const newAttrs = {};
    resource.resourceAttrs.forEach(attr => (newAttrs[attr.key] = attr));
    const existingAttrs = {};
    attributes.forEach(attr => (existingAttrs[attr.key] = attr));
    const resultAttrs = values<Attribute>(assign({}, existingAttrs, newAttrs));

    return resultAttrs
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).addState(
          resource.key,
          resultAttrs.map(attr => {
            if (attr.key === 'version') {
              let version = parseInt(attr.value as string, 10);
              return {
                type: '1',
                key: 'version',
                value: `${++version}`,
                immutable: false
              };
            }
            return attr;
          })
        )
      : null;
  },
  addPolicy: async policy =>
    policy
      ? stateList<Policy>(NS.POLICY, context).addState(policy.key, policy)
      : null,
  deleteMSPAttrByMSPID: async mspid =>
    mspid
      ? stateList(NS.MSP_ATTRIBUTE, context).deleteStateByKey(makeKey([mspid]))
      : null,
  deletePolicyById: async id =>
    id
      ? stateList(NS.POLICY, context).deleteStatesByKeyRange([
          JSON.stringify(id)
        ])
      : null,
  deletePolicyByIdSid: async (id, sid) =>
    id && sid
      ? stateList(NS.POLICY, context).deleteStateByKey(makeKey([id, sid]))
      : null,
  deleteReourceAttrByURI: async uri =>
    uri
      ? stateList(NS.RESOURCE_ATTRIBUTE, context).deleteStateByKey(
          makeKey(uri.split('/'))
        )
      : null,
  getMSPAttrByMSPID: async mspid =>
    mspid
      ? stateList<Attribute[]>(NS.MSP_ATTRIBUTE, context).getState(
          makeKey([mspid])
        )
      : null,
  getPolicyById: async id =>
    id
      ? stateList<Policy>(NS.POLICY, context).getQueryResult([
          JSON.stringify(id)
        ])
      : null,
  getPolicyByIdSid: async (id, sid) =>
    id && sid
      ? stateList<Policy>(NS.POLICY, context).getState(makeKey([id, sid]))
      : null,
  getResourceAttrGroupByURI: async uri =>
    uri
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).getQueryResult(
          uri.split('/').map(part => JSON.stringify(part))
        )
      : null,
  getResourceAttrByURI: async uri =>
    uri
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).getState(
          makeKey(uri.split('/'))
        )
      : null
});
