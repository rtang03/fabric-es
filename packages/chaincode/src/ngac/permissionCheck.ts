import { Context } from 'fabric-contract-api';
import { isEqual } from 'lodash';
import { ngacRepo } from './ngacRepo';
import { policyDecisionEngine } from './policyDecisionEngine';
import { postAssertion } from './postAssertion';
import { Assertion, Attribute, CONTEXT, NAMESPACE, RESOURCE } from './types';
import { createId, createResource } from './utils';

enum FCN {
  CREATE_COMMIT = 'createCommit',
  QUERY_BY_ENTITYID = 'queryByEntityId',
  QUERY_BY_ENTITYID_COMMITID = 'queryByEntityIdCommitId',
  DELETE_BY_ENTITYID_COMMITID = 'deleteByEntityIdCommitId',
  DELETE_BY_ENTITYID = 'deleteByEntityId',
  ADD_MSP_ATTR = 'addMSPAttr',
  ADD_RESORUCE_ATTR = 'addResourceAttr',
  ADD_POLICY = 'addPolicy',
  DELETE_MSP_ATTR_BY_MSPID = 'deleteMSPAttrByMSPID',
  DELETE_POLICY_BY_ID = 'deletePolicyById',
  DELETE_POLICY_BY_ID_SID = 'deletePolicyByIdSid',
  DELETE_RESOURCE_ATTR_BY_URI = 'deleteReourceAttrByURI',
  UPSERT_RESOURCE_ATTR = 'upsertResourceAttr',
  GET_MSP_ATTR_BY_MSPID = 'getMSPAttrByMSPID',
  GET_POLICY_BY_ID = 'getPolicyById',
  GET_POLICY_BY_ID_SID = 'getPolicyByIdSid',
  GET_RESOURCE_ATTR_BY_URI = 'getResourceAttrByURI'
}

const noPolicyRequired = Promise.resolve([
  { sid: 'system', message: 'No policy required', assertion: true }
]);

export const permissionCheck: (option: {
  policyClass?: string;
  context: Context;
}) => Promise<Assertion[]> = async ({ context, policyClass }) => {
  const { stub, clientIdentity } = context;
  const { fcn, params } = stub.getFunctionAndParameters();
  const type: any = '1';
  const attr = (key, value, immutable = true) =>
    value ? { type, key, value, immutable } : null;
  const mspAttrs = await ngacRepo(context).getMSPAttrByMSPID(
    clientIdentity.getMSPID()
  );
  let resourceAttrs: Attribute[] = [];
  const cn = clientIdentity.getX509Certificate().subject.commonName;
  const mspid = clientIdentity.getMSPID();
  const id = createId([mspid, cn]);
  return {
    [FCN.CREATE_COMMIT]: async (): Promise<Assertion[]> => {
      const [entityName, entityId, version, evenStr] = params;
      resourceAttrs = [
        attr(CONTEXT.INVOKER_ID, id),
        attr(CONTEXT.INVOKER_MSPID, mspid),
        attr(CONTEXT.SUBJECT_CN, cn)
      ].filter(item => !!item);

      if (version === '0') {
        [
          attr(RESOURCE.VERSION, version, false),
          attr(RESOURCE.ENTITYNAME, entityName),
          attr(RESOURCE.ENTITYID, entityId),
          attr(RESOURCE.CREATOR_MSPID, mspid),
          attr(RESOURCE.CREATOR_CN, cn),
          attr(RESOURCE.CREATOR_ID, id)
        ]
          .filter(item => !!item)
          .forEach(item => resourceAttrs.push(item));
      } else {
        const uri = `${NAMESPACE.MODEL}/${mspid}/${entityName}/${entityId}`;
        const attrs = await ngacRepo(context).getResourceAttrByURI(uri);

        if (isEqual(attrs, [])) {
          return [
            { sid: 'system', assertion: false, message: 'No resource found' }
          ];
        } else attrs.forEach(item => resourceAttrs.push(item));
      }

      const eventsJson: any[] = JSON.parse(evenStr);
      const eventTypes = eventsJson.map(({ type }) => type);
      const policies = await ngacRepo(context).getPolicyById(id);
      const target = createResource({
        context,
        entityName,
        entityId,
        resourceAttrs,
        mspAttrs
      });
      return isEqual(policies, [])
        ? [{ sid: 'system', assertion: false, message: 'No policy found' }]
        : policyDecisionEngine(policies, context)
            .request({ eventTypes, target })
            .then(async assertions => {
              await postAssertion(assertions, target, context);
              return assertions;
            });
    },
    [FCN.QUERY_BY_ENTITYID]: () => noPolicyRequired,
    [FCN.QUERY_BY_ENTITYID_COMMITID]: () => noPolicyRequired,
    [FCN.DELETE_BY_ENTITYID]: () => noPolicyRequired,
    [FCN.DELETE_BY_ENTITYID_COMMITID]: () => noPolicyRequired,
    [FCN.ADD_MSP_ATTR]: () => noPolicyRequired,
    [FCN.ADD_RESORUCE_ATTR]: () => noPolicyRequired,
    [FCN.ADD_POLICY]: () => noPolicyRequired,
    [FCN.DELETE_POLICY_BY_ID]: () => noPolicyRequired,
    [FCN.DELETE_POLICY_BY_ID_SID]: () => noPolicyRequired,
    [FCN.DELETE_RESOURCE_ATTR_BY_URI]: () => noPolicyRequired,
    [FCN.DELETE_MSP_ATTR_BY_MSPID]: () => noPolicyRequired,
    [FCN.GET_POLICY_BY_ID]: () => noPolicyRequired,
    [FCN.GET_POLICY_BY_ID_SID]: () => noPolicyRequired,
    [FCN.GET_RESOURCE_ATTR_BY_URI]: () => noPolicyRequired,
    [FCN.GET_MSP_ATTR_BY_MSPID]: () => noPolicyRequired,
    [FCN.UPSERT_RESOURCE_ATTR]: () => noPolicyRequired
  }[fcn]() as Promise<Assertion[]>;
};
