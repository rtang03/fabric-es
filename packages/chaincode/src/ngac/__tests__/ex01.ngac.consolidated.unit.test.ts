import { ngacRepo } from '../ngacRepo';
import { permissionCheck } from '../permissionCheck';
import { CONTEXT, NAMESPACE as NS, RESOURCE, RESOURCE as RES } from '../types';
import { createId, createPolicy, createResource } from '../utils';
import { createMSPResource } from '../utils/createMSPResource';
import { mspAttributeDb, policyDb, resourceAttributeDb } from './__utils__';

jest.mock('../ledger-api/statelist');

const context: any = {
  stub: {
    getFunctionAndParameters: jest.fn()
  },
  clientIdentity: {
    getMSPID: jest.fn(),
    getID: jest.fn(),
    getX509Certificate: jest.fn()
  },
  resourceAttributeDb,
  mspAttributeDb,
  policyDb
};
let entityName;
let entityId;
let version;
let eventStr;
let id;
context.clientIdentity.getMSPID.mockImplementation(() => 'Org3MSP');
context.clientIdentity.getX509Certificate.mockImplementation(() => ({
  subject: { commonName: 'Tester01@example.com' },
  issuer: { commonName: 'rca-org3' }
}));

describe('Example 1: PolicyEngine/CRUD Tests', () => {
  beforeEach(() => {
    entityName = 'dev_subject';
    entityId = 'subject_01';
    version = '0';
    id = 'x509::/O=Dev/OU=client/CN=Tester01@example.com';
    context.clientIdentity.getID.mockImplementation(() => id);
  });

  it('should create Org3MSP attribute', async () => {
    const mspId = 'Org3MSP';
    const mspAttrs = [
      { type: '1', key: 'mspid', value: 'Org3MSP' },
      { type: '1', key: 'env', value: 'test' }
    ];
    const mspResource = createMSPResource({ context, mspId, mspAttrs });
    await ngacRepo(context)
      .addMSPAttr(mspResource)
      .then(attrs => expect(attrs).toEqual(mspAttrs));
  });

  it('should create entityName attribute', async () => {
    const resourceAttrs = [
      { type: '1', key: 'entityNameOwner', value: 'tester' },
      {
        type: 'N',
        key: 'createSubject',
        value: [createId(['Org3MSP', 'Tester01@example.com'])]
      }
    ];
    await ngacRepo(context)
      .addResourceAttr(
        createResource({
          context,
          entityName,
          resourceAttrs
        })
      )
      .then(attrs => expect(attrs).toEqual(resourceAttrs));
  });

  it('should create policy', async () => {
    const policy = createPolicy({
      context,
      policyClass: 'consolidated-test',
      sid: 'allowCreateSubject',
      allowedEvents: ['SubjectCreated'],
      uri: `${NS.MODEL}/${NS.ORG}?id=resourceAttrs:${RES.CREATOR_MSPID}/${NS.ENTITY}?id=resourceAttrs:${RES.ENTITYNAME}`,
      condition: {
        hasList: { createSubject: `${RES.CREATOR_ID}` }
      },
      effect: 'Allow'
    });

    await ngacRepo(context)
      .addPolicy(policy)
      .then(policy => expect(policy).toMatchSnapshot());
  });

  it('should createSubject', async () => {
    eventStr = JSON.stringify([{ type: 'SubjectCreated' }]);
    context.stub.getFunctionAndParameters.mockImplementation(() => ({
      fcn: 'createCommit',
      params: [entityName, entityId, version, eventStr]
    }));

    return permissionCheck({ context }).then(assertions =>
      expect(assertions).toEqual([{ sid: 'allowCreateSubject', assertion: true }])
    );
  });

  // note that the mocked implementation is based on unit test;
  // it invokes permissionCheck directly, and skip the EventStore Contract
  // it can use version = '0'. The corresponding integration test cannot
  // repeat the same logic, coz re-create version '0' will fail.
  it('should fail createNothing', async () => {
    eventStr = JSON.stringify([{ type: 'NothingCreated' }]);
    context.stub.getFunctionAndParameters.mockImplementation(() => ({
      fcn: 'createCommit',
      params: [entityName, entityId, version, eventStr]
    }));
    return permissionCheck({ context }).then(assertions =>
      expect(assertions).toEqual([{ sid: 'system', assertion: false, message: 'No policy found' }])
    );
  });

  it('should create policy #2 with both hasList and strictEquals', async () => {
    const policy = createPolicy({
      context,
      policyClass: 'consolidated-test',
      sid: 'allowUpdateSubject',
      allowedEvents: ['SubjectUpdated'],
      uri: `${NS.MODEL}/${NS.ORG}?id=resourceAttrs:${RES.CREATOR_MSPID}/${NS.ENTITY}?id=resourceAttrs:${RES.ENTITYNAME}/${NS.ENTITYID}?id=resourceAttrs:${RES.ENTITYID}`,
      condition: {
        hasList: { updateSubject: RES.CREATOR_ID },
        stringEquals: { [CONTEXT.INVOKER_MSPID]: RESOURCE.CREATOR_MSPID }
      },
      effect: 'Allow'
    });

    await ngacRepo(context)
      .addPolicy(policy)
      .then(({ sid }) => expect(sid).toEqual(policy.sid));
  });

  it('should add resourceAttribute of updateSubject', async () => {
    const resourceAttrs = [
      {
        type: 'N',
        key: 'updateSubject',
        value: [createId(['Org3MSP', 'Tester01@example.com'])]
      }
    ];
    await ngacRepo(context)
      .upsertResourceAttr(
        createResource({
          context,
          entityName,
          entityId,
          resourceAttrs
        })
      )
      .then(attributes => attributes.filter(({ key }) => key === 'updateSubject'))
      .then(attribute =>
        expect(attribute).toEqual([
          {
            type: 'N',
            key: 'updateSubject',
            value: [createId(['Org3MSP', 'Tester01@example.com'])]
          }
        ])
      );
  });

  it('should updateSubject', async () => {
    version = '1';
    eventStr = JSON.stringify([{ type: 'SubjectUpdated' }]);
    context.stub.getFunctionAndParameters.mockImplementation(() => ({
      fcn: 'createCommit',
      params: [entityName, entityId, version, eventStr]
    }));

    return permissionCheck({ context }).then(assertions =>
      expect(assertions).toEqual([{ sid: 'allowUpdateSubject', assertion: true }])
    );
  });
});
