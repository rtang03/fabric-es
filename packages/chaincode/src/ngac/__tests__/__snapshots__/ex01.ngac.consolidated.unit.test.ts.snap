// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example 1: PolicyEngine/CRUD Tests should create policy 1`] = `
Object {
  "allowedEvents": Array [
    "SubjectCreated",
  ],
  "attributes": Object {
    "uri": "model/org?id=resourceAttrs:creator_mspid/entity?id=resourceAttrs:entityName",
  },
  "condition": Object {
    "hasList": Object {
      "createSubject": "creator_id",
    },
  },
  "effect": "Allow",
  "key": "\\"Org3MSP::Tester01@example.com\\"~\\"allowCreateSubject\\"",
  "policyClass": "consolidated-test",
  "sid": "allowCreateSubject",
}
`;
